crates/nu-protocol/src/value/record.rs

```diff
@@ -11,6 +15,137 @@ pub struct Record {
     inner: Vec<(String, Value)>,
 }
 
+/// A wrapper around [`Record`] that handles lookups. Whether the keys are compared case sensitively
+/// or not is controlled with the `SENSITIVE` parameter.
+///
+/// It is never actually constructed as a value and only used as a reference to an existing
+/// [`Record`].
+#[repr(transparent)]
+pub struct CaseTypedRecord<const SENSITIVE: bool>(Record);
+
+impl<const SENSITIVE: bool> CaseTypedRecord<SENSITIVE> {
+    #[inline]
+    const fn from_record(record: &Record) -> &Self {
+        // SAFETY: `CaseTypedRecord` has the same memory layout as `Record`.
+        unsafe { &*(record as *const Record as *const Self) }
+    }
+
+    #[inline]
+    const fn from_record_mut(record: &mut Record) -> &mut Self {
+        // SAFETY: `CaseTypedRecord` has the same memory layout as `Record`.
+        unsafe { &mut *(record as *mut Record as *mut Self) }
+    }
+
+    #[inline]
+    fn keys_eq(lhs: &str, rhs: &str) -> bool {
+        match SENSITIVE {
+            true => lhs == rhs,
+            false => lhs.eq_ignore_case(rhs),
+        }
+    }
+
+    pub fn index_of(&self, col: impl AsRef<str>) -> Option<usize> {
+        let col = col.as_ref();
+        self.0.columns().rposition(|k| Self::keys_eq(k, col))
+    }
+
+    pub fn contains(&self, col: impl AsRef<str>) -> bool {
+        self.index_of(col.as_ref()).is_some()
+    }
+
+    pub fn get(&self, col: impl AsRef<str>) -> Option<&Value> {
+        let index = self.index_of(col.as_ref())?;
+        Some(self.0.get_index(index)?.1)
+    }
+
+    pub fn get_mut(&mut self, col: impl AsRef<str>) -> Option<&mut Value> {
+        let index = self.index_of(col.as_ref())?;
+        Some(self.0.get_index_mut(index)?.1)
+    }
+
+    /// Remove single value by key and return it
+    pub fn remove(&mut self, col: impl AsRef<str>) -> Option<Value> {
+        let index = self.index_of(col.as_ref())?;
+        Some(self.0.remove_index(index))
+    }
+
+    /// Insert into the record, replacing preexisting value if found.
```

Comment:
This functionality is really an upsert vs insert. I'm not sure it matters for the api but I was confused when I read it. I see below that this terminology was already used before and you're just tweaking things.

Reply:
I think `insert` is correct here. Because this displaces the original value. My gut feeling about an `upsert` would be that it only inserts the value if there isn't one already.

Reply:
upsert means update a value if it already exists otherwise insert a new key/value. i could live with this api being either. I'm just saying that I think insert is not tecnically the right word based on the comments.